File: src/com/oscarparty/servlets/PicksSubmittedServlet.java
package com.oscarparty.servlets;

import com.oscarparty.servlets.playerpicks.CategoryPicks;
import com.oscarparty.servlets.playerpicks.PlayerPicks;
import com.oscarparty.servlets.playerpicks.PlayerPicksDAO;
import scala.collection.mutable.DefaultEntry;
import scala.collection.mutable.HashMap;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Arrays;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

/**
 * Servlet for when a player has submitted picks to be saved.
 */
public class PicksSubmittedServlet extends HttpServlet {

    static {
        try {
            Class.forName("org.postgresql.Driver");
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
    }

    public void doPost(HttpServletRequest req, HttpServletResponse res)
            throws ServletException, IOException {

        Map<String, String[]> parameterMap = req.getParameterMap();
        Set<Map.Entry<String, String[]>> paramEntries = parameterMap.entrySet();
        PlayerPicks playerPicks = new PlayerPicks();
        PrintWriter writer = res.getWriter();
        writer.println("Your picks have been submitted!");
        for (Map.Entry<String, String[]> paramKey : paramEntries) {
            String pickKey = paramKey.getKey();
            String[] pickValues = paramKey.getValue();
            writer.println("Parameter key: " + pickKey + " with value " +
                    Arrays.toString(pickValues));

            if (pickValues.length > 1) {
                throw new ServletException("More pick values than allowed.");
            } else if (pickValues.length == 1) {
                String pick = pickValues[0];
                if (pickKey.equals("userName")) {
                    playerPicks.userName_$eq(pick);
                } else {
                    //result params are like "Best Actress in a Supporting Role.botPick" = "[Amy Adams]"
                    int indexOfDot = pickKey.indexOf(".");
                    if (indexOfDot < 0) throw new ServletException("Pick key is malformed: " + pickKey);
                    String categoryName = pickKey.substring(0, indexOfDot);
                    String pickPriority = pickKey.substring(indexOfDot + 1);
                    playerPicks.addPick(categoryName, pickPriority, pick);
                }
            }
        }
        if (!playerPicks.arePicksValid()) {
            throw new IllegalArgumentException("Picks were not valid!");
        }
        PlayerPicksDAO playerPicksDAO = new PlayerPicksDAO();
        playerPicksDAO.storePicks(playerPicks);

        RequestDispatcher dispatcher = getServletContext().getRequestDispatcher("/picksSubmitted.jsp");
        dispatcher.forward(req, res);
    }

}

File: src/com/oscarparty/servlets/PicksSubmittedServlet.scala
//package com.oscarparty.servlets
//
//import java.io.PrintWriter
//import java.util.{Map => JMap}
//import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
//import javax.servlet.{RequestDispatcher, ServletException}
//
//import com.oscarparty.servlets.PicksSubmittedServlet.EachPickIdentifier
//import com.oscarparty.servlets.data.PlayerPicksDAO
//import com.oscarparty.servlets.data.PlayerPicksDAO.PlayerPickForInsertion
//import com.oscarparty.servlets.data.nominees.OscarNomineesDAO
//
//import scala.collection.JavaConverters._
//
///**
// * Servlet for when a player has submitted picks to be saved.
// */
//class PicksSubmittedServlet extends HttpServlet {
//
//  override def doPost(req: HttpServletRequest, res: HttpServletResponse) {
//    res.setContentType("text/html; charset=UTF-8")
//    res.setCharacterEncoding("UTF-8")
//
//    val parameterMapJava: JMap[String, Array[String]] = req.getParameterMap
//    val parameterMap = parameterMapJava.asScala.toMap
//
//    val writer: PrintWriter = res.getWriter
//
//    //find the newly added player
//    val newPlayer = findPlayerFromRequest(parameterMap)
//    val playerPicks = filterPicksFromRequest(parameterMap)
//
//    val convertedPlayerPicks = convertPlayerPicks(playerPicks)
//
//    PlayerPicksDAO.addPlayerPicks(newPlayer, convertedPlayerPicks)
//    writer.println("Your picks have been submitted!")
//    val dispatcher: RequestDispatcher = getServletContext.getRequestDispatcher("/picksSubmitted.jsp")
//    dispatcher.forward(req, res)
//  }
//
//  private def convertPlayerPicks(playerPicks: Map[EachPickIdentifier, Int]): List[PlayerPickForInsertion] = {
//    //iterate through the categories, finding the picks for each rank and create the PlayerPick object
//
//    def findPick(categoryId: Int, rankName: String): Int = {
//      val nomineeNamePicked = playerPicks(EachPickIdentifier(categoryId, rankName))
//      OscarNomineesDAO.getNominee(nomineeNamePicked).id
//    }
//
//    OscarNomineesDAO.getCategories.map { category =>
//      PlayerPickForInsertion(category.id,
//        findPick(category.id, "topPick"),
//        findPick(category.id, "midPick"),
//        findPick(category.id, "botPick"))
//    }.toList
//  }
//
//  private def findPlayerFromRequest(requestParams: Map[String, Array[String]]): String = {
//    val newPlayerEntry = requestParams.iterator.find { isPlayerNameParam }
//    newPlayerEntry match {
//      case Some((key, values)) =>
//        assertOneValue(key, values)
//        //the first value is the player name
//        values.head
//      case _ => throw new ServletException("No player name added")
//    }
//  }
//
//  //TODO refactor to get the ID of the nominees and categories instead of the names. Have to thread the IDs through to the form on the page.
//  private def filterPicksFromRequest(requestParams: Map[String, Array[String]]): Map[EachPickIdentifier, Int] = {
//    requestParams.filter(!isPlayerNameParam(_)).map { case (key, selectionList) =>
//      //result params are like <id>.<rank> = [<nominee>] aka "2.botPick" = "[Amy Adams]"
//      assertOneValue(key, selectionList)
//      val splitCategory = key.split('.')
//      if (splitCategory.length != 2)
//        throw new ServletException(s"Request param key $key does not make sense, no '.' separator for category name and pick rank.")
//      EachPickIdentifier(Integer.parseInt(splitCategory(0)), splitCategory(1)) -> Integer.parseInt(selectionList.head)
//    }
//  }
//
//  private def isPlayerNameParam(paramEntry: (String, Array[String])): Boolean = {
//    paramEntry match {
//      case (key, values) => key == "userName"
//      case _ => false
//    }
//  }
//
//  def assertOneValue(key: String, values: Array[String]) {
//    if (values.length != 1) {
//      throw new ServletException(s"There were the wrong number of values ($values) for $key")
//    }
//  }
//}
//
//
//object PicksSubmittedServlet {
//  private case class EachPickIdentifier(categoryId: Int, pickRank: String)
//}